### PR TITLE
Fix DocType closing bracket recognition in buffered reader

### LINKED ISSUE #533: Deserialization of a doctype with very long content fails
`quick_xml::de::from_reader()` parsing fails if the XML contains a doctype with content larger than the internal `BufRead` capacity.
For instance
```xml
<!DOCTYPE [
<!-- A very very long comment *snipped* -->
]>
<X></X>
```

Here is a minimal code to reproduce this issue. It fails with an `ExpectedStart` error.
```rust
use std::io::Write;
use serde::Deserialize;

#[derive(Deserialize)]
struct X {}

fn main() {
    {
        let mut file = std::fs::File::create("test.xml").unwrap();
        let header = &"<!DOCTYPE X [<!--";
        let footer = &"-->]><X></X>";
        let padding = 8192 - (header.len() + 2);
        write!(file, "{header}{:1$}{footer}", "", padding).unwrap();
    }

    let file = std::fs::File::open("test.xml").unwrap();
    let reader = std::io::BufReader::new(file);
    let _: X = quick_xml::de::from_reader(reader).unwrap();
}
```
`Cargo.toml` content
```toml
[package]
name = "test"
version = "0.1.0"
edition = "2021"

[dependencies]
quick-xml = { version = "0.27.1", features = ["serialize"] }
serde = { version = "1.0", features = ["derive"] }
```

* When decreasing the padding size, or using `BufReader::with_capacity()` to increase the buffer, even of 1 byte, there is no error.
* Other `BufRead` implementations don't have this issue (checked with `&[u8]` and `stdin`).
* Content does not have to be in one "block". The same issue occurs for a doctype split into multiple declarations and comments.
* With a longer doctype with real content, the error may be different. For instance it may complains about an invalid `!` from a `!ENTITY` tag.
* No issue with `serde-xml-rs`, even for larger comments.
* Tested on Windows, with rustc 1.66.0.

### LINKED ISSUE #590: Unexpected Bang Since 0.23.1
The following file can't be parsed since 0.23.1 as it throws an UnexpectedBang(69) when parsing with the given example code. Simply ignoring the error causes the given example code not to parse anything at all. Version 0.22.0 worked fine
[JMdict.zip](https://github.com/tafia/quick-xml/files/11202365/JMdict.zip)
This is just a trimmed file of the JMdict file to demonstrate the issue. You can find the original file [here](http://ftp.edrdg.org/pub/Nihongo/JMdict.gz).

Working v0.22.0 Code:
```rust
use quick_xml::{events::Event, Reader};
use std::{fs::File, io::BufReader};

fn main() {
    let mut reader = Reader::from_reader(BufReader::new(File::open("./JMdict_short").unwrap()));
    reader.trim_text(true);
    let mut buf = vec![];

    loop {
        match reader.read_event(&mut buf) {
            Ok(Event::Start(ref e)) => {
                if let b"entry" = e.name() {
                    println!("{}", std::str::from_utf8(e.name()).unwrap());
                }
            }

            Ok(Event::Eof) => break,

            Err(e) => {
                panic!("{e:?}");
            }

            _ => (),
        }
    }
}
```

Broken v0.28.1 Code (same code, just migrated):
```rust
use quick_xml::{events::Event, Reader};
use std::{fs::File, io::BufReader};

fn main() {
    let mut reader = Reader::from_reader(BufReader::new(File::open("./JMdict_short").unwrap()));
    reader.trim_text(true);
    let mut buf = vec![];

    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(ref e)) => {
                if let b"entry" = e.name().0 {
                    println!("{}", std::str::from_utf8(e.name().0).unwrap());
                }
            }

            Ok(Event::Eof) => break,

            Err(e) => {
                panic!("{e:?}");
            }

            _ => (),
        }
    }
}
```

### LINKED ISSUE #801: Error when parsing !DOCTYPE longer than buffer of BufRead
## Description
When parsing `<!DOCTYPE []>` from a BufRead, there is a bug where it thinks a closing tag `'>'` that occurs as part of the doctype instead closes the DOCTYPE. This occurs when the entire doctype does not fit inside the bufread's buffer.

Because of that, when a closing tag '>' is mistakenly parsed as end of doctype, the content after that is parsed incorrectly.

## Steps to reproduce
```rust
use quick_xml::events::Event;
use quick_xml::Reader;
use std::io::BufReader;

fn main() {
    let xml = r#"<!DOCTYPE X [<!-- comment --><!ENTITY a "a">]>"#;
    let reader = BufReader::with_capacity(4, xml.as_bytes());
    let mut xml_reader = Reader::from_reader(reader);
    let mut buf = vec![];
    loop {
        match xml_reader.read_event_into(&mut buf).unwrap() {
            Event::Eof => break,
            _ => {}
        }
    }
    println!("Parsed xml without error");
}
```

Above code throws an error `Syntax(InvalidBangMarkup)`. This is because the closing tag for comment `<!-- comment -->` is mistakenly parsed as closing doctype, and so trying to parse `<!ENTITY>` throws an error.

However, things work fine if we increase the bufreader capacity to greater than text content (e.g. 1024).

I'll be opening a PR for the fix right away

### PR DESCRIPTION
**Pull request by BlueGreenMagick**
Fixes #533, fixes #590, fixes #801

### CODE REVIEW

**Comment by BlueGreenMagick** [comment]
Modified the code so that `balance_buf` is computed only once.

**Review by Mingun — CHANGES_REQUESTED** [review]
That PR will not fix the problem entirely, because correct DTD can contain unbalanced number of `<` and `>`, but I think that is rare case, so until proper fix will be implemented, a partial one also acceptable. The implementation, however, still may be improved. The problem here that we do not store state between invocations of `BangType::parse`. So the solution is to store balance inside `BangType::DocType`, then, I think, it will be unnecessary to calculate it over `buf` (because everything in `buf` previously was in `chunk`).

I also see the problem in the external `for i in memchr::memchr_iter(b'>', chunk)` loop. It will skip `<`s inside `chunk`s that does not contain `>`, so the balance will be calculated incorrectly. I think, if create a test carefully, it will catch such situation.

So I ask you to do the following changes:
- convert `BangType::DocType` -> `BangType::DocType(i32)`
- start with balance == 1 (it is `<` in `<!DOCTYPE`) here
https://github.com/tafia/quick-xml/blob/2f3824a1d265ac0b683aa485683ad00ddbc5c092/src/reader/mod.rs#L1031
- when calculate balance take into account saved balance
- emit event only when `BangType::DocType(0)` here
https://github.com/tafia/quick-xml/blob/2f3824a1d265ac0b683aa485683ad00ddbc5c092/src/reader/state.rs#L147
- add tests for the mentioned case (chunk with `<` + chunk with `>` which closes `<` from the first chunk). It is enough just craft such a string and use appropriate buffer size to slice string to desired chunks
- also add regression test from https://github.com/tafia/quick-xml/issues/590#issuecomment-1503927845
- try to check if this fixes #533. I think it should because it seems to be a duplicate of these issues, but need to check

**Comment by BlueGreenMagick** [comment]
Thanks for the detailed feedback! I adjusted the PR.

> * start with balance == 1 (it is `<` in `<!DOCTYPE`) here
As the balance is calculated in up to chunk excluding current `>`, 1 is not added to balance initially.

> add tests for the mentioned case (chunk with < + chunk with > which closes < from the first chunk). It is enough just craft such a string and use appropriate buffer size to slice string to desired chunks

Are regression tests for issues 533, 590, 801 enough? Or are additional tests desired?

File: tests/issues.rs [hunk 1]
```diff
@@ -171,6 +171,40 @@ mod issue514 {
...
+    #[derive(Deserialize)]
+    struct X {}
+
+    #[test]
+    fn test() {
+        let xml = "<!DOCTYPE X [<!-- -->]><X></X>";
+        let reader = BufReader::with_capacity(20, xml.as_bytes());
+        let _: X = quick_xml::de::from_reader(reader).unwrap();
+    }
+}
```
Comment (Mingun) [inline tests/issues.rs:188]: Please put this test into `serde-issues.rs`. It compiles only when `serialize` feature is active so does not require feature gate.

File: src/reader/mod.rs [hunk 2]
```diff
@@ -1020,15 +1020,15 @@ enum BangType {
     /// <!--...-->
     Comment,
     /// <!DOCTYPE...>
-    DocType,
+    DocType(i32),
```
Comment (Mingun) [inline src/reader/mod.rs:1023]: Updating documentation also would be useful (language can be improved if required, I'm not native speaker):
```rust
    /// <!DOCTYPE...>. Contains a balance of `<` (+1) and `>` (-1)
    DocType(i32),
```

**Review by Mingun — APPROVED** [review]
> Are regression tests for issues 533, 590, 801 enough? Or are additional tests desired?

I'm not sure that tests still cover the case when we get four chunks with roughly such content:
1. `<!DOCTYPE`
2. `<` without any `>`
3. `>`, which would close `<` from point 2, not the doctype
4. `>` which should close doctype

If that case already covered, fine, although I would prefer to have test which explicitly guaranties such conditions (can tweak one of added regression tests for that).

And please add changelog entry to `Changelog.md` that people know that bug (#590) fixed. I'll close other as duplicate.

**Comment by Mingun** [comment]
You also can squash all your changes if you wish

**Comment by BlueGreenMagick** [comment]
Done! 

I adjusted regression test for issue801 so that all angle brackets are explicitly in different buffer (by lowering buffer size to 2 bytes).

I think it might be better (and easier with GitHub UI) to squash merge on your end, so others can follow the conversations and commits in the future.

**Comment by Mingun** [comment]
Thanks!

**Pull request merged**